File: Cpp/Best Time to Buy and Sell Stock.cpp
#include <vector>

class Solution {
public:
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    int maxProfit(vector<int> &prices) {
        // write your code here
        int len = prices.size();
        if (len <= 1) return 0;
        int cur = prices[0], maxp = 0;
        for (int i = 1; i < len; i++) {
            cur = min(cur, prices[i]);
            maxp = max(maxp, prices[i] - cur);
        }
        return maxp;
    }
};

File: Java/Best-Time-to-Buy-and-Sell-Stock.java
public class Solution {
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    public int maxProfit(int[] prices) {
        // write your code here
        int maxP = 0, minPrice = Integer.MAX_VALUE;
        for (int i = 0; i < prices.length; i ++) {
            maxP = Math.max(maxP, prices[i] - minPrice);
            minPrice = Math.min(minPrice, prices[i]);
        }
        return maxP;
    }
}
